postgres_composite_types/__init__.py

```diff
@@ -43,71 +43,76 @@
 from django.db.backends.signals import connection_created
 from django.dispatch import Signal
 from psycopg2 import ProgrammingError
-from psycopg2.extras import register_composite
+from psycopg2.extensions import AsIs, ISQLQuote, adapt, register_adapter
+from psycopg2.extras import CompositeCaster, register_composite
 
 LOGGER = logging.getLogger(__name__)
 
 __all__ = ['CompositeType']
 
 
-class BaseField(models.Field):
-    """Base class for the field that relates to this type."""
-
-    Meta = None
-
-    def db_type(self, connection):
-        LOGGER.debug("db_type")
+class QuotedCompositeType(object):
+    """
+    A wrapper for CompositeTypes that knows how to convert itself into a safe
+    postgres representation. Created from CompositeType.__conform__
+    """
+    prepared_value = None
 
-        if not isinstance(connection, PostgresDatabaseWrapper):
-            raise RuntimeError("Composite types are only available "
-                               "for postgres")
+    def __init__(self, obj):
+        self.obj = obj
+        self.model = obj._meta.model
 
-        return self.Meta.db_type
+        self.values = [
+            adapt(field.get_db_prep_value(field.value_from_object(self.obj),
+                                          self.model.registered_connection))
+            for _, field in self.model._meta.fields]
 
-    def to_python(self, value):
-        LOGGER.debug("to_python: > %s", value)
+    def __conform__(self, protocol):
+        # Required for nested composite types
+        if protocol is ISQLQuote:
+            return self
```

Comment:
Generally some comments explaining where you can see how this implementation works would be useful for the next hapless reader trying to understand how the frack this works.